src/niri.rs

```diff
@@ -3728,47 +3736,162 @@ impl Niri {
-                        &output,
-                        screencopy.overlay_cursor(),
-                        RenderTarget::ScreenCapture,
-                    )
-                    .into_iter()
-                    .rev();
-
-                let region_loc = screencopy.region_loc();
-                let elements = elements.map(|element| {
-                    RelocateRenderElement::from_element(
-                        element,
-                        region_loc.upscale(-1),
-                        Relocate::Relative,
-                    )
-                });
+        self.layout.update_render_elements(output);
 
-                let scale = output.current_scale().fractional_scale().into();
-                let transform = output.current_transform();
-                render_to_shm(renderer, screencopy.buffer(), scale, transform, elements)
-                    .context("error rendering to screencopy shm buffer: {err:?}")?;
+        let elements = self.render(
+            renderer,
+            output,
+            screencopy.overlay_cursor(),
+            RenderTarget::ScreenCapture,
+        );
+        let Some(queue) = self.screencopy_state.get_queue_mut(manager) else {
+            trace!("screencopy manager destroyed already");
+            return Ok(());
+        };
+        let damage_tracker = queue.split().0;
+
+        let render_result = Self::render_for_screencopy_internal(
+            renderer,
+            output,
+            &elements,
+            screencopy.with_damage(),
+            damage_tracker,
+            &screencopy,
+        )
+        .map(|_| ());
+
+        match render_result {
+            Ok(()) => screencopy.submit(false),
+            Err(_) => {
+                // recreate damage tracker to report full damage next check.
+                *damage_tracker = OutputDamageTracker::new((0, 0), 1.0, Transform::Normal);
+            }
+        }
+        render_result
+    }
+
+    fn render_for_screencopy_internal<'a>(
+        renderer: &mut GlesRenderer,
+        output: &Output,
+        elements: &[OutputRenderElements<GlesRenderer>],
+        with_damage: bool,
+        damage_tracker: &'a mut OutputDamageTracker,
+        screencopy: &Screencopy,
+    ) -> anyhow::Result<Option<&'a Vec<Rectangle<i32, Physical>>>> {
+        let OutputModeSource::Static {
+            size: last_size,
+            scale: last_scale,
+            transform: last_transform,
+        } = damage_tracker.mode().clone()
+        else {
+            unreachable!("output always has static mode");
+        };
 
-                screencopy.submit(false);
+        let size = screencopy.buffer_size();
+        let scale: Scale<f64> = output.current_scale().fractional_scale().into();
+        let transform = output.current_transform();
+
+        if size != last_size || scale != last_scale || transform != last_transform {
+            *damage_tracker = OutputDamageTracker::new(size, scale, transform);
```

Comment:
As far as I understand, this will always use the screencopy buffer size for rendering. However, when an output resizes after copy_with_damage(), it should fail the screencopy frame, instead of using an incorrect size for rendering.

This makes me realize that after capture_output_region() there's no way to tell if an output resized or not from the sizes alone. So, perhaps, after all, we should just fail all screencopies for the output from `Niri::output_resized()`?

Reply:
I don't think it is specified what to do when an output changes size. Right now it is just reporting full damage. We can save the output size at the time of request in the info and then compare it and only fail if it has changed.

Reply:
Right now, as far as I read what this code will do, is it will render the output with the wrong size (the size from screencopy, whereas the real output size is already different).

Reply:
Yes it is still rendering at the initial size. I meant the doc is not clear on what to do in this case. What do you propose?

Reply:
In `Niri::output_resized()` if the size changed, fail the screencopies (maybe also if transform/scale changed but idk)

Reply:
Indeed I think it should fail if any of the three changes since whats rendered in screencopy is now different from what's displayed.